simpleviewinc/keg-hub
# Zen 352 retheme updates - Part 1

**Ticket**: [ZEN-352](https://jira.simpleviewtools.com/browse/ZEN-352)

### Notes
* This is `Part 1` of the changes being made. The changes were starting to build up, so I figured this is a good break-point
* All parts are fully functional, and should work as expected
* Some parts need some adjustments, which will come in the next pull request
* There is a matching [tap-events-force](https://github.com/simpleviewinc/tap-events-force/pull/46) pull request, will be coming shortly

## Context/Goal
* Events force is defining their styles in CSS
* Their clients can define custom CSS as well
* Currently, the tap, using re-theme, is only compatible with JSON style objects
* We need to be able to pull styles from stylesheets already applied to the DOM and apply them to keg-components

## Updates
**Keg-Core**
* `repos/keg-core/core/configs/webpack.config.js`
  * Update to pull the web export of re-theme
  * Webpack is only used in a web context, so updating the webpack build will not effect native
**Re-Theme**
* `rollup`
  * Updated dependencies in `package.json` for rollup
  * Updated `.babelrc` to be `babel.config.js`
  * Updated to export platform to separate folders
  * Split the export into multiple exports
    * Not all apps will need the `styles-parser` or the `ReThemeHeadProvider`
    * These can be exported separately
    * All other exports should work same as before
* Removed old builds and folder structure
  * folder structure now looks like this
    * `build/<output type>/<platform>/*`
      * output type = `esm` || `cjs`
      * platform = `web` || `native`
* Added tests for a number of files
  * Still more to add, which will come in the next pull request
* `src/hooks/useCss.js` - new hook
  * Dynamically builds the data-class attribute, and pulls the styles from the theme
  * This is all based off the passed in `theme-reference`, See `Theme Reference` section below for more details
* `src/styleParser/styleSheetParser.js`
  * Accepts an array of `class names`
  * Parses `styles sheets` found on the dom and looks for matching `class-names`
  * Any found will be passed to a callback, or use the default callback
  * Default callback converts the matching `class-names` into `data-class attributes`
* `src/theme/restructureTheme.js`
  * Updated to extracts the `$className` || `$class` from the theme
    * Normalizes it to be just `$class`
    * Adds it to the merged theme after platforms have been merged
* Rename `context` and `provider` to `reThemeContext` and `reThemeProvider` for clarity
  * They are no longer the only context and provider with the Head code added
* Added `src/head/*`
  * Creates a new `context` and `provider` to allow attaching styles to a Dom Stylesheet
  * Allows using css `sudo` classes among other things


## Theme Reference && useCss
* A new attribute (`$class` || `$className`) can be added to the theme object to define the `root-class-name` for a component
* When used with the `useCss` hook we can do a number of things:
  * Attach the styles to the Dom Header attribute instead of inline styles
    * Which allows us to use `css sudo` classes like `:hover` and `:active`
  * Dynamically build the `data-class=*` attribute based on the root class
    * The `dataSet` no longer needs to be defined within the Component
    * The `data-class=*` is dynamically built for a components based on the `themeRef.$class` attribute of the theme
      * The `$class` can also be overwritten with a passed in `id`
  * Can be used with the `styleSheetParser` helper to merge all styles together
    * This will allow for custom `Events-Force` styles to be merged in with our Theme
 ### Workflow
  * On App Init, we use `styleSheetParser` to pull matching `classes` from the dom
    * This will return a `CssInJs` object that can be merged in with our theme
  * With in the theme the `$class` key is added, which should match the classes returned from the `styleSheetParser`
    * This then allows us to map the parsed `styles` from the dom with the theme styles where they should be applied
  * In the components where needed, use the `useCss` hook to build the `data-class` for the components, and apply the styles to the dom
  * We now have a merged theme with custom styles, and with out loss of functionality

### Example
  * Code looks like this => http://snpy.in/psqWfU
    * In code see `tap-events-force/src/components/SessionTime/sessionTime.js`
  * Dom looks like this => http://snpy.in/mmuNiG

## Testing
* Docker Package url => `keg evf pack run docker.pkg.github.com/simpleviewinc/keg-packages/tap:zen-352-css-updates`
  * Ensure the style tag gets added to the dom `head` element, see image above for how it should look
  * Ensure the styles are applied to the `ef-session-date-time-content-text` data class => http://snpy.in/ybSQ3k

* Run tests for the `Tap` and `ReTheme`
  * To Run Tap tests
    * `keg evf pack run docker.pkg.github.com/simpleviewinc/keg-packages/tap:zen-352-css-updates test`
      * **Ensure all tests pass**
  * To Run Re-Theme tests
    * Pull this branch `keg retheme && keg pr 8` 
    * `yarn install && yarn test`
      * **Ensure all tests pass**
